Update the constant-pH tutorial

kosovan opened the pull request:
Fixes #3941

Description of changes:
- add prior knowledge, make equations and units more consistent
- add a flag to run electrostatics with Debye-Hückel instead of P3M
- accelerate tutorial with DH, cell system, non-interacting particle type
- fix ambiguous explanations

kosovan committed:
First changes to the cpH tutorial

kosovan committed:
reduced units and unit conversions now hopefully consistent

kosovan committed:
Various changes to cpH tutorial

prob_reaction, more on Bjerrum and thermodynamics. Use easier-to-read
nested cycles for non-bonded interactions between all particle pairs.

kosovan committed:
Various small cleanups before working on the performance

kosovan committed:
Improve performance of the cpH tutorial

Use Debye-Hückel electrostatics, specify a non_interacting_type
and apply other tricks to speed-up the calculation.

kosovan commented:
Some interesting observations on speedup:
* `RE.set_non_interacting_type(len(TYPES))` speeds up the calculation of interacting system roughly by a factor of 2.0.
*  Replacing P3M with DH makes the calculation with electrostatics almost as fast as without
*  If we do reaction in every cycle and integration just in some cycles, we save the amount of runtime roughly proportional to `PROB_INTEGRATION`. In the previous version we were doing the opposite, so we were wasting some 50% of the reaction samples.

If the non_interacting_type has such an effect on the speed, I think it is desired to give it a fixed and low value, or to set dynamically to the lowest possible value that is available.

Maybe we should consider using some of these tricks also in the tests of the RxMC and cpH algorithms. It is not important to get perfectly correct physics but to ensure that we can still reproduce the titration curve of an interacting system if we change something in the algorithm. What do you think of that? @jngrad @RudolfWeeber

jngrad commented:
> Maybe we should consider using some of these tricks also in the tests of the RxMC and cpH algorithms.

Good idea, I can have a look at this later this week to speed up the testsuite, and mention it in the docs too.

A couple of tests did not pass CI. Let us know when you are ready with the PR and we'll review it.

davidbbeyer commented on doc/tutorials/constant_pH/constant_pH.ipynb:
```diff
@@ -4,7 +4,20 @@
    "cell_type": "markdown",
```
<p>In the first paragraph I would add "particle size \sigma <strong>in nm </strong>..."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4262/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='11'/>

kosovan replied:
@davidbbeyer I have not yet cleaned up the text. Modifications to the python code are mostly done but I only did those changes in the text that were directly related to changes in the code.

kosovan replied:
<p>@davidbeyer: I do not understand, where would you like to add "particle size \sigma <strong>in nm"</strong> and why? In the current implementation, you can set particle size in any unit of length and pint takes care of converting it.</p>

davidbbeyer replied:
<p>I'm okay with leaving it out, but then you should also remove the reference to Kelvin ("We start by setting the desired temperature&nbsp;T in Kelvin") since it is an arbitrary unit as well.</p>

kosovan replied:
<p>OK, now I see the point. Right, we should remove the Kelvin from there.</p>

davidbbeyer commented on doc/tutorials/constant_pH/constant_pH.ipynb:
```diff
@@ -4,7 +4,20 @@
    "cell_type": "markdown",
```
<p>I think we should change the beginning of some of the sentences, it is not so nice to have successive sentences starting with "If" or "We".</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4262/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

kosovan replied:
<p>OK, I can rephrase it. However, I do not see it as an important issue.</p>

davidbbeyer commented on doc/tutorials/constant_pH/constant_pH.ipynb:
```diff
@@ -4,7 +4,20 @@
    "cell_type": "markdown",
```
<p>There is a typo here: "tne number ..." instead of "the number ..."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4262/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='19'/>

davidbbeyer replied:
<p>I think it would be better to write this as a list like this:</p><ul><li>time-step ...</li><li>skin ...</li><li>seed ...</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/espressomd/espresso/pull/4262/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='25'/>

kosovan committed:
Cleanup of the text of the constant-pH tutorial

kosovan committed:
Resolved some comments from David and some typos

kosovan commented:
TODO: before this PR is merged, we should change the initial set of parameters back to the ideal system and clear the cells with solutions of the excercises.

jngrad committed:
Proofread cpH tutorial

Properly stylize "pH" and "pKa" throughout the text.
Fix plot labels, typos and formatting. Address TODOs.
Make citations link to the bibliographic entries and
cleanup links to the user guide. Run the tutorial
without electrostatics to reproduce the ideal case.

jngrad committed:
tests: Adapt cpH tutorial tests

jngrad committed:
Merge branch 'python' into cph-tutorial-cleanup

jngrad reviewed (approved):
Thanks for the improvements!
